### PR TITLE
Improve qwen vl impl

### PR DESCRIPTION
**Pull request by drbh**
This PR improves qwen2-vl in the following ways

- [X] re enable and improve test (remove `max_tokens`)
- [X] adds estimation logic for vision model flops and consume vision config in launcher
   - this improves the startup input token limit (avoids defaulting to 4096 for all vlms)
- [X] removes cuda graphs <=3 for qwen-vl in the launcher
- [X] add impl for `RotaryPositionEmbeddingMultimodalSections` in `rotary.py`
- [X] slice position ids on subsequent cuda graph warmup calls to avoid slicing in rotary (attempting to remove now)

### CODE REVIEW

**Comment by drbh** [comment]
This PR improves the performance and response from qwen2 vl based models. Small reproducible examples can be run with the startup commands and script below

### expected output

```bash
text-generation-launcher --model-id bytedance-research/UI-TARS-7B-DPO
```

```json
{
    "generated_text": "The image depicts the Statue of Liberty, a renowned landmark located on Liberty Island in New York Bay."
}
{
    "generated_text": "The image features the logo of Flash Attention, a state-of-the-art attention mechanism designed for transformers,"
}
{
    "generated_text": "The image features a stylized illustration of a rabbit, rendered in a minimalist and abstract design. The"
}
```

```bash
text-generation-launcher --model-id Qwen/Qwen2-VL-2B-Instruct
```

```json
{
    "generated_text": "The image depicts the iconic Statue of Liberty in New York City, with the city's skyline in the"
}
{
    "generated_text": "The image compares two different implementations of attention mechanisms in neural networks:\n\n### Standard Attention Implementation"
}
{
    "generated_text": "The image depicts a rabbit in an astronaut's suit standing on a rocky, red-brown planet with"
}
```

```bash
text-generation-launcher --model-id Qwen/Qwen2-VL-7B-Instruct --num-shard 2
```

```json
{
    "generated_text": "The image depicts the iconic Statue of Liberty, a colossal neoclassical sculpture on Liberty Island in"
}
{
    "generated_text": "The image compares the standard attention implementation with Flash Attention in the context of memory and computation operations.\n\n###"
}
{
    "generated_text": "The image depicts an astronaut in a futuristic space suit standing on a rocky surface with a reddish-orange"
}
```

script for testing a couple images with the models

```python
import requests
import json

url = "http://127.0.0.1:3000/generate"

headers = {"Content-Type": "application/json"}

image_urls = [
    "https://cdn.britannica.com/61/93061-050-99147DCE/Statue-of-Liberty-Island-New-York-Bay.jpg",
    "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/tgi/flash-attn.png",
    "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/tgi/rabbit.png",
]

for image in image_urls:
    query = "Describe the image"

    payload = {
        "inputs": f"<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\n![]({image}){query}<|im_end|>\n<|im_start|>assistant\n",
        "parameters": {"max_new_tokens": 20},
    }

    response = requests.post(url, headers=headers, json=payload)

    # print the response
    print(json.dumps(response.json(), indent=4))
```

File: server/text_generation_server/models/__init__.py [hunk 1]
```diff
@@ -1248,7 +1248,7 @@ def get_model(
             revision=revision,
             quantize=quantize,
             speculator=speculator,
-            dtype=dtype,
+            dtype=torch.bfloat16,
```
Comment (danieldk) [inline server/text_generation_server/models/__init__.py:1251]: Should be passed through with `default_dtype` I think?

Comment (drbh) [reply server/text_generation_server/models/__init__.py:1251]: ah good catch! updating to use `default_dtype` in the latest commit

File: server/text_generation_server/layers/rotary.py [hunk 2]
```diff
@@ -86,6 +86,16 @@ def static(cls, config, dim, base, device):
             # `rope_type` is now standard in transformers, but some existing models
             # have `type` instead.
             rope_type = rope_scaling.get("rope_type", rope_scaling.get("type", None))
+            mrope_section = rope_scaling.get("mrope_section", None)
+
+            # only apply mrope if sections are provided and the rope type is mrope or default
+            if mrope_section is not None and (
+                rope_type == "mrope" or rope_type == "default"
```
[the remark this answers is not in the document]
Comment (drbh) [reply server/text_generation_server/layers/rotary.py:93]: yea makes sense I've updated to only match if `rope_type == "mrope"` and ensure the `rope_type` is set to `mrope` within the qwen2-vl model. 

note: currently transformers `AutoModel` returns `rope_type` default for all qwen-vl models

File: launcher/src/main.rs [hunk 3]
```diff
@@ -277,8 +279,50 @@ impl Config {
...
+                tracing::debug!(
+                    "Vision flops: {}",
+                    human_size(vision_flops as usize, "flop")
+                );
+                Some(text_flops + vision_flops)
+            }
+            // model has a vision config but is not supported for flops calculation
+            // we return None to avoid overestimating the memory requirements
+            _ => None,
+        }
```
[the remark this answers is not in the document]
Comment (drbh) [reply launcher/src/main.rs]: make sense, will followup with another PR that avoid the specific model

File: server/text_generation_server/layers/rotary.py [hunk 4]
```diff
@@ -548,3 +558,76 @@ def apply_llama3_scaling(
...
+        split_cos, split_sin = [
+            torch.split(t, self.sections, dim=-1) for t in (cos, sin)
+        ]
+        cos = torch.cat([m[i % 3] for i, m in enumerate(split_cos)], dim=-1).unsqueeze(
+            1
+        )
+        sin = torch.cat([m[i % 3] for i, m in enumerate(split_sin)], dim=-1).unsqueeze(
+            1
+        )
+        # prepare input tensors
```
[the remark this answers is not in the document]
Comment (drbh) [reply server/text_generation_server/layers/rotary.py]: agreed, and moved into `get_cos_sin` in the latest commit

File: server/text_generation_server/models/flash_causal_lm.py [hunk 5]
```diff
@@ -1400,7 +1400,11 @@ def cuda_graph_warmup(self, bs: int, max_s: int, max_bt: int):
         cache_lengths = [0] * bs
         if max_bs is None:
             input_ids = torch.zeros(bs, dtype=torch.int64, device=self.device)
-            position_ids = torch.zeros(bs, dtype=torch.int32, device=self.device)
+            if hasattr(self.model, "get_position_ids"):
+                # use model specific position ids for initialization
+                position_ids = self.model.get_position_ids(input_ids)
```
[the remark this answers is not in the document]
Comment (drbh) [reply server/text_generation_server/models/flash_causal_lm.py]: currently the position ids are not mutated by the model before they are placed on the batch, which is why this logic lives outside of the model. 

I've improved the logic a bit to expand `position_ids` if rope_scaling exists and the type is mrope, in this case the position ids are expanded to the size of the sections. 

This isn't a huge change from the original code but feel a bit cleaner. Please let me know if theres a better solution 🙏

File: server/text_generation_server/models/flash_causal_lm.py [hunk 6]
```diff
@@ -1427,7 +1431,7 @@ def cuda_graph_warmup(self, bs: int, max_s: int, max_bt: int):
                     "Cuda graphs should be generated in decreasing order size to reduce VRAM usage"
                 )
             input_ids = self.cuda_graphs[max_bs]["input_ids"][:bs]
-            position_ids = self.cuda_graphs[max_bs]["position_ids"][:bs]
+            position_ids = self.cuda_graphs[max_bs]["position_ids"][..., :bs]
```
[the remark this answers is not in the document]
Comment (drbh) [reply server/text_generation_server/models/flash_causal_lm.py]: yea agreed, i've refactored the position ids to always have the batch in the first dim and fixed the indexing to slice correctly (which also fixes all issues with cuda graphs)

File: server/text_generation_server/layers/rotary.py [hunk 7]
```diff
@@ -548,3 +558,76 @@ def apply_llama3_scaling(
...
+        q, k = [x.transpose(0, 1).unsqueeze(0) for x in (query, key)]
+        rotary_dim = cos.shape[-1]
+        q1, k1 = q[..., :rotary_dim], k[..., :rotary_dim]
+        q2 = torch.cat((-q[..., rotary_dim // 2 :], q[..., : rotary_dim // 2]), dim=-1)
+        k2 = torch.cat((-k[..., rotary_dim // 2 :], k[..., : rotary_dim // 2]), dim=-1)
+
+        rotary_emb.apply_rotary(q1, q2, cos, sin, q1, q2, True)
+        rotary_emb.apply_rotary(k1, k2, cos, sin, k1, k2, True)
```
[the remark this answers is not in the document]
Comment (drbh) [reply server/text_generation_server/layers/rotary.py]: good point, I've simplified the code to avoid the transpose and simply rotate q and k and pass them to `rotary_emb.apply_rotary` in the latest commits

File: server/text_generation_server/layers/rotary.py [hunk 8]
```diff
@@ -86,11 +86,24 @@ def static(cls, config, dim, base, device):
             # `rope_type` is now standard in transformers, but some existing models
             # have `type` instead.
             rope_type = rope_scaling.get("rope_type", rope_scaling.get("type", None))
+            mrope_section = rope_scaling.get("mrope_section", None)
+
+            # only apply mrope if sections are provided and the rope type is mrope and a section is provided
+            if mrope_section is not None and rope_type == "mrope":
+                mrope_section = rope_scaling.get("mrope_section")
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/layers/rotary.py:93]: I think this can be moved down to the `rope_type == "mrope":` block below?

File: server/text_generation_server/layers/rotary.py [hunk 9]
```diff
@@ -548,3 +561,70 @@ def apply_llama3_scaling(
             new_freqs.append((1 - smooth) * freq / scaling_factor + smooth * freq)
 
     return torch.tensor(new_freqs, dtype=freqs.dtype, device=freqs.device)
+
+
+class RotaryPositionEmbeddingMultimodalSections(PositionRotaryEmbedding):
+    def __init__(self, inv_freq, scaling_factor, sections):
```
Comment (danieldk) [inline server/text_generation_server/layers/rotary.py:567]: NIT: Types

File: server/text_generation_server/layers/rotary.py [hunk 10]
```diff
@@ -548,3 +561,70 @@ def apply_llama3_scaling(
...
+        # rotate half the sequence length
+        rot = cos.shape[-1] // 2
+        q2 = torch.cat([-query[..., rot:], query[..., :rot]], dim=-1)
+        k2 = torch.cat([-key[..., rot:], key[..., :rot]], dim=-1)
+
+        # apply the rotation
+        rotary_emb.apply_rotary(query, q2, cos, sin, query, q2, True)
+        rotary_emb.apply_rotary(key, k2, cos, sin, key, k2, True)
+
+    def _update_cos_sin_cache(self, dtype, device, seqlen):
```
Comment (danieldk) [inline server/text_generation_server/layers/rotary.py:589]: NIT: types

File: server/text_generation_server/layers/rotary.py [hunk 11]
```diff
@@ -548,3 +561,70 @@ def apply_llama3_scaling(
...
+        position_ids: torch.Tensor,
+        max_s: int,
+        dtype: torch.dtype,
+    ):
+        self._update_cos_sin_cache(dtype, position_ids.device, max_s)
+
+        # access freqs for each of the 3 sections and stack them
+        cos_c = torch.stack(
+            [self._cos_cached[position_ids[:, i]] for i in range(3)], dim=0
+        )
```
Comment (danieldk) [inline server/text_generation_server/layers/rotary.py:614]: I wonder if the looping/stacking is necessary? I think we can do this with a single gather?

Comment (danieldk) [reply server/text_generation_server/layers/rotary.py:614]: Also, is this reliably 3? Or should it be the length of `mrope_section`?

File: server/text_generation_server/layers/rotary.py [hunk 12]
```diff
@@ -548,3 +561,70 @@ def apply_llama3_scaling(
...
+        sin_c = torch.stack(
+            [self._sin_cached[position_ids[:, i]] for i in range(3)], dim=0
+        )
+
+        # chunk based on sections
+        split_cos = torch.split(cos_c, self.sections, dim=-1)
+        split_sin = torch.split(sin_c, self.sections, dim=-1)
```
Comment (danieldk) [inline server/text_generation_server/layers/rotary.py:621]: How do the sections here relate to the sections above?

File: server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py [hunk 13]
```diff
@@ -251,7 +235,8 @@ def forward(
         max_s,
         prefill_cache_indices,
     ):
-        normed_hidden_states, res = self.input_layernorm(hidden_states, residual)
+        residual = hidden_states
+        normed_hidden_states, _ = self.input_layernorm(hidden_states)
```
Comment (danieldk) [inline server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py:239]:
```suggestion
        normed_hidden_states, residual = self.input_layernorm(hidden_states)
```

should be the same as long as no residual is passed?

https://github.com/huggingface/text-generation-inference/blob/88fd56f549637b36529582b682c4c76eb78e36a5/server/text_generation_server/layers/layernorm.py#L44

File: server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py [hunk 14]
```diff
@@ -266,15 +251,14 @@ def forward(
...
         # faster post attention rms norm
-        normed_attn_res_output, attn_res = self.post_attention_layernorm(
-            attn_output, res
-        )
-
-        mlp_output = self.mlp(normed_attn_res_output)
-
-        return mlp_output, attn_res
+        residual = hidden_states
+        hidden_states, _ = self.post_attention_layernorm(hidden_states)
```
Comment (danieldk) [inline server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py:258]: Same here

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py [hunk 15]
```diff
@@ -222,12 +222,11 @@ def __init__(self, prefix, config, weights):
     def forward(
         self, hidden_states, cu_seqlens, rotary_pos_emb, max_seqlen
     ) -> torch.Tensor:
-        hidden_states_post_norm1, res = self.norm1(hidden_states)
-        hidden_states = hidden_states + self.attn(
-            hidden_states_post_norm1, cu_seqlens, rotary_pos_emb, max_seqlen
-        )
-        hidden_states_post_norm2, res = self.norm2(hidden_states)
-        hidden_states = hidden_states + self.mlp(hidden_states_post_norm2)
+        norm1_out, _ = self.norm1(hidden_states)
```
Comment (danieldk) [inline server/text_generation_server/models/custom_modeling/qwen2_vl.py:225]: See above.

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py [hunk 16]
```diff
@@ -409,97 +412,91 @@ def __init__(self, prefix, config, weights):
...
-                    llm_pos_ids_list[-1].max() + 1 if len(llm_pos_ids_list) > 0 else 0
-                )
-                text_len = batch_input_ids.size(1) - current_pos
-                llm_pos_ids_list.append(
-                    torch.arange(text_len, device=d).view(1, -1).expand(3, -1) + st_idx
-                )
+        input_ids: torch.Tensor,
+        image_grid_thw: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        if image_grid_thw is None:
```
Comment (danieldk) [inline server/text_generation_server/models/custom_modeling/qwen2_vl.py:418]: Is there similar code in transformers? Would be nice to have a reference here.

File: server/text_generation_server/layers/rotary.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (drbh) [reply server/text_generation_server/layers/rotary.py]: oh thanks this is a duplicated, removed in the latest commit

File: server/text_generation_server/layers/rotary.py (hunk 9 above)
Comment (drbh) [reply server/text_generation_server/layers/rotary.py]: 🙏 updated in the latest commit

File: server/text_generation_server/layers/rotary.py (hunk 10 above)
Comment (drbh) [reply server/text_generation_server/layers/rotary.py]: 🙏 updated in the latest commit

File: server/text_generation_server/layers/rotary.py (hunk 11 above)
Comment (drbh) [reply server/text_generation_server/layers/rotary.py]: agreed! I reworked the logic to avoid the loop/stacking and now cache a tensor of the indices thats used to gather in `get_cos_sin`. This approach removed a lot of the extra reshaping and limits the initialization. Thanks!

File: server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py (hunk 13 above)
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py]: oh good catch, i've updated to avoid the extra variable/line and confirmed the outputs match the logits in the transformers implementation

File: server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py (hunk 14 above)
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py]: thanks! updated

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py (hunk 15 above)
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/qwen2_vl.py]: updated too

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py (hunk 16 above)
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/qwen2_vl.py:429]: yes, theres a slightly different implementation here https://github.com/huggingface/transformers/blob/e284c7e954abe12c34b50461c17f8115a0afe115/src/transformers/models/qwen2_vl/modeling_qwen2_vl.py#L1391 however the implementation in this PR is slightly shorter and relies on torch indexing when possible.

I've added a link to the reference and improved the comments to be more readable in the latest commit

Comment (danieldk) [reply server/text_generation_server/models/custom_modeling/qwen2_vl.py:429]: Thanks!

**Review by danieldk — APPROVED** [review]
Thanks for all the changes, looks good to me! Since this is a fairly large change, it would be good if someone else can do a final sanity check (@Narsil ?).

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py [hunk 17]
```diff
@@ -527,6 +520,7 @@ def forward(
 
         # apply the visual model to the pixel values if they are provided
         if pixel_values is not None and len(pixel_values) > 0:
+            pixel_values = pixel_values.to(inputs_embeds.dtype)
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/qwen2_vl.py:523]: Why is this necessary ?

**Review by Narsil — APPROVED** [review]
LGTM.

This is much better !
